drivers/i2c/i2c_imx.c

```diff
@@ -0,0 +1,451 @@
+
+	/* Wait for the transfer to complete */
+	k_sem_take(&data->device_sync_sem, K_FOREVER);
+
+	return transfer->ack;
+}
+
+static void i2c_imx_read(struct device *dev, u8_t *rxBuffer, u8_t rxSize)
+{
+	I2C_Type *base = DEV_BASE(dev);
+	struct i2c_imx_data *data = DEV_DATA(dev);
+	struct i2c_master_transfer *transfer = &data->transfer;
+
+	transfer->isBusy = true;
+
+	/* Clear I2C interrupt flag to avoid spurious interrupt */
+	I2C_ClearStatusFlag(base, i2cStatusInterrupt);
+
+	/* Change to receive state. */
+	I2C_SetDirMode(base, i2cDirectionReceive);
+	transfer->currentDir = i2cDirectionReceive;
+
+	transfer->rxBuff = rxBuffer;
+	transfer->rxSize = rxSize;
+
+	if (transfer->rxSize == 1) {
+		/* Send Nack */
+		I2C_SetAckBit(base, false);
+	} else {
+		/* Send Ack */
+		I2C_SetAckBit(base, true);
+	}
+
+	/* dummy read to clock in 1st byte */
+	I2C_ReadByte(base);
+
+	/* Enable I2C interrupt, subsequent data transfer will be handled
+	 * in ISR.
+	 */
+	I2C_SetIntCmd(base, true);
+
+	/* Wait for the transfer to complete */
+	k_sem_take(&data->device_sync_sem, K_FOREVER);
+
+}
+
+static int i2c_imx_configure(struct device *dev, u32_t dev_config_raw)
+{
+	I2C_Type *base = DEV_BASE(dev);
+	struct i2c_imx_data *data = DEV_DATA(dev);
+	struct i2c_master_transfer *transfer = &data->transfer;
+	u32_t baudrate;
+
+	if (!(I2C_MODE_MASTER & dev_config_raw)) {
+		return -EINVAL;
+	}
+
+	if (I2C_ADDR_10_BITS & dev_config_raw) {
+		return -EINVAL;
+	}
+
+	/* Initialize I2C state structure content. */
+	transfer->txBuff = 0;
+	transfer->rxBuff = 0;
+	transfer->cmdSize = 0;
+	transfer->txSize = 0;
+	transfer->rxSize = 0;
+	transfer->isBusy = false;
+	transfer->currentDir = i2cDirectionReceive;
+	transfer->currentMode = i2cModeSlave;
+
+	switch (I2C_SPEED_GET(dev_config_raw)) {
+	case I2C_SPEED_STANDARD:
+		baudrate = KHZ(100);
+		break;
+	case I2C_SPEED_FAST:
+		baudrate = MHZ(1);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Setup I2C init structure. */
+	i2c_init_config_t i2cInitConfig = {
+		.baudRate	  = baudrate,
+		.slaveAddress = 0x00
+	};
+
+	i2cInitConfig.clockRate = get_i2c_clock_freq(base);
+
+	I2C_Init(base, &i2cInitConfig);
+
+	I2C_Enable(base);
+
+	return 0;
+}
+
+static int i2c_imx_transfer(struct device *dev, struct i2c_msg *msgs,
+		u8_t num_msgs, u16_t addr)
+{
+	I2C_Type *base = DEV_BASE(dev);
+	struct i2c_imx_data *data = DEV_DATA(dev);
+	struct i2c_master_transfer *transfer = &data->transfer;
+	u8_t *buf, *buf_end;
+	unsigned int flags;
+	int result = -EIO;
+
+	if (!num_msgs) {
+		return 0;
+	}
+
+	/* We want an initial Start condition */
+	flags = I2C_MSG_RESTART;
+
+	/* Make sure we're in a good state so slave recognises the Start */
+	I2C_SetWorkMode(base, i2cModeSlave);
+	transfer->currentMode = i2cModeSlave;
+	/* Switch back to Rx direction. */
+	I2C_SetDirMode(base, i2cDirectionReceive);
+	transfer->currentDir = i2cDirectionReceive;
+
+	do {
+
+		/* Start condition? */
+		if (flags & I2C_MSG_RESTART) {
```

Comment:
This condition will only be satisfied on the first iteration and doesn't need to be inside the do/while loop. If you move it out, then you can move `flags |= msgs->flags` to the beginning of the loop, and use `flags` consistently throughout the loop. The current mix of `flags` and `msgs->flags` is confusing.

Reply:
Done